File: diff/comparison/histogram_matcher.go
package comparison

import (
	"github.com/byron1st/file-diff/diff/histogram"
)

// HistogramMatcher implements LineMatcher using the Histogram Diff algorithm.
// It extends Patience Diff by using line occurrence frequency to select
// anchors, handling repetitive structures (JSON, YAML, boilerplate) where
// Patience falls back to Myers due to lack of unique lines.
type HistogramMatcher struct{}

func (m *HistogramMatcher) Match(left, right []string, policy ComparisonPolicy) FairDiffIterable {
	normLeft := normalizeLines(left, policy)
	normRight := normalizeLines(right, policy)

	matches := histogram.Diff(normLeft, normRight, func(l, r []string) []histogram.Anchor {
		return histogramMyersFallback(l, r)
	})

	builder := NewChangeBuilder(len(left), len(right))
	for _, match := range matches {
		builder.MarkEqual(match.LeftIdx, match.RightIdx)
	}
	return Fair(builder.Finish())
}

// histogramMyersFallback uses CompareLines (Myers) to produce matched pairs.
func histogramMyersFallback(left, right []string) []histogram.Anchor {
	if len(left) == 0 || len(right) == 0 {
		return nil
	}
	result := CompareLines(left, right, PolicyDefault)
	var matches []histogram.Anchor
	for _, u := range result.Unchanged() {
		count := u.End1 - u.Start1
		for i := range count {
			matches = append(matches, histogram.Anchor{LeftIdx: u.Start1 + i, RightIdx: u.Start2 + i})
		}
	}
	return matches
}

File: swift/Sources/FileDiff/HistogramMatcher.swift
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Swift (via the Go port in this repository).

/// Line matcher backed by the Histogram Diff algorithm.
///
/// Extends Patience Diff by using line occurrence frequency to select
/// anchors, which handles repetitive structures (JSON, YAML, boilerplate)
/// where Patience would fall back to Myers.
public struct HistogramMatcher: LineMatcher {
    public init() {}

    public func match(_ left: [String], _ right: [String], policy: ComparisonPolicy) -> FairDiffIterable {
        let normLeft = normalizeLines(left, policy: policy)
        let normRight = normalizeLines(right, policy: policy)

        let matches = histogramDiff(normLeft, normRight) { subLeft, subRight in
            histogramMyersFallback(subLeft, subRight)
        }

        let builder = ChangeBuilder(length1: left.count, length2: right.count)
        for match in matches {
            builder.markEqual(match.leftIndex, match.rightIndex)
        }
        return fair(builder.finish())
    }
}

func histogramMyersFallback(_ left: [String], _ right: [String]) -> [HistogramAnchor] {
    if left.isEmpty || right.isEmpty {
        return []
    }
    let result = compareLines(left, right, policy: .default)
    var matches: [HistogramAnchor] = []
    for unchanged in result.unchanged {
        let count = unchanged.left.count
        for i in 0..<count {
            matches.append(HistogramAnchor(
                leftIndex: unchanged.left.lowerBound + i,
                rightIndex: unchanged.right.lowerBound + i
            ))
        }
    }
    return matches
}
